### PR TITLE
Fix legacy RC QP indexing for NVSHMEM 3.5.19+

### PR DESCRIPTION
**Pull request by caoxiaoyuyuyuyuyu**
## Background

The legacy IBGDA path currently indexes `globalmem.rcs` as PE-major. That matches NVSHMEM 3.4.5 and earlier, but NVSHMEM changed RC QP allocation to QP-major, PE-interleaved order in NVIDIA/nvshmem@ce9d4872522775c50ee536a7044bc36c688db2b9. The change is included in NVSHMEM 3.5.19 and later.

With NVSHMEM 3.7.2, the old expression can select another PE RC QP. All ranks may initialize successfully, then the first cross-node dispatch stops making progress and can eventually surface a CUDA fault.

## Change

- Calculate the queue number once.
- Use `queue * npes + pe` for NVSHMEM 3.5.19 and later.
- Preserve `pe * queues_per_pe + queue` for older supported NVSHMEM releases.
- Select the device layout at compile time using NVSHMEM vendor version macros.
- Query the loaded host-library version before NVSHMEM initialization and reject compile/runtime combinations that cross the 3.5.19 layout transition. This closes the compatibility gap left by the normal NVSHMEM check, which permits an older device module with a newer host library.
- Fail compilation if the required vendor version macros are unavailable instead of silently selecting the old layout.

## Validation

- Built the first current-main change (`d268280`) successfully for SM 10.3 with CUDA 13, NVSHMEM 3.7.2, and NCCL 2.30.7; `cuobjdump` confirms `arch = sm_103`.
- Verified from official NVSHMEM 3.4.5, 3.5.19, and 3.7.2 headers that `nvshmemx_vendor_get_version_info` and all three vendor version macros are available.
- Added compile-time boundary assertions for 3.5.18 (old layout) and 3.5.19 (new layout); the same classifier also passed a local C++17 boundary check.
- Applied the equivalent device-index fix to DeepEP v1.2.1 built for SM 10.3 with CUDA 13 and NVSHMEM 3.7.2.
- Passed multi-node EP32 differential checks for dispatch/combine forward output, input gradients, and route-probability gradients.
- Passed two simultaneous EP32 groups and ten consecutive finite full-model optimization steps without a dispatch timeout, CUDA illegal-access error, NCCL error, process restart, or Xid.
- The longer CP1 diagnostic stopped before step 10 on a model-loss `torch.OutOfMemoryError` while materializing FP32 vocabulary logits. The failure was outside DeepEP/NVSHMEM; no transport or device-fault signature preceded it.
- Formatted both touched files with clang-format 15.0.7.

### CODE REVIEW

File: csrc/kernels/legacy/ibgda_device.cuh [hunk 1]
```diff
@@ -80,9 +79,16 @@ __device__ static __forceinline__ nvshmemi_ibgda_device_state_t* ibgda_get_state
...
     auto state = ibgda_get_state();
-    const auto num_rc_per_pe = ibgda_get_state()->num_rc_per_pe;
-    return &state->globalmem
-                .rcs[pe * num_rc_per_pe * state->num_devices_initialized + id % (num_rc_per_pe * state->num_devices_initialized)];
+    const auto num_rcs = state->num_rc_per_pe * state->num_devices_initialized;
+    const auto qp = id % num_rcs;
+#if NVSHMEM_VENDOR_MAJOR_VERSION > 3 || (NVSHMEM_VENDOR_MAJOR_VERSION == 3 && NVSHMEM_VENDOR_MINOR_VERSION > 5) || \
+    (NVSHMEM_VENDOR_MAJOR_VERSION == 3 && NVSHMEM_VENDOR_MINOR_VERSION == 5 && NVSHMEM_VENDOR_PATCH_VERSION >= 19)
+    // Since 3.5.19, NVSHMEM stores RC QPs in QP-major, PE-interleaved order.
+    return &state->globalmem.rcs[qp * nvshmemi_device_state_d.npes + pe];
```
[the remark this answers is not in the document]
Comment (caoxiaoyuyuyuyuyu) [reply csrc/kernels/legacy/ibgda_device.cuh]: Adjusted in 2c70ee1 by naming the common-state value once as `num_pes`. The IBGDA transport state does not contain `npes` in the supported NVSHMEM headers, so the value still comes from `nvshmemi_device_state_d` while the RC pointer and queue counts come from the IBGDA state.

File: csrc/kernels/legacy/ibgda_device.cuh [hunk 2]
```diff
@@ -80,9 +79,16 @@ __device__ static __forceinline__ nvshmemi_ibgda_device_state_t* ibgda_get_state
 
 __device__ static __forceinline__ nvshmemi_ibgda_device_qp_t* ibgda_get_rc(int pe, int id) {
     auto state = ibgda_get_state();
-    const auto num_rc_per_pe = ibgda_get_state()->num_rc_per_pe;
-    return &state->globalmem
-                .rcs[pe * num_rc_per_pe * state->num_devices_initialized + id % (num_rc_per_pe * state->num_devices_initialized)];
+    const auto num_rcs = state->num_rc_per_pe * state->num_devices_initialized;
+    const auto qp = id % num_rcs;
```
[the remark this answers is not in the document]
Comment (caoxiaoyuyuyuyuyu) [reply csrc/kernels/legacy/ibgda_device.cuh:87]: Addressed in 2c70ee1. Both touched translation units now fail compilation unless all three `NVSHMEM_VENDOR_{MAJOR,MINOR,PATCH}_VERSION` macros are defined.

File: csrc/kernels/legacy/ibgda_device.cuh [hunk 3]
```diff
@@ -80,9 +79,16 @@ __device__ static __forceinline__ nvshmemi_ibgda_device_state_t* ibgda_get_state
...
+#if NVSHMEM_VENDOR_MAJOR_VERSION > 3 || (NVSHMEM_VENDOR_MAJOR_VERSION == 3 && NVSHMEM_VENDOR_MINOR_VERSION > 5) || \
+    (NVSHMEM_VENDOR_MAJOR_VERSION == 3 && NVSHMEM_VENDOR_MINOR_VERSION == 5 && NVSHMEM_VENDOR_PATCH_VERSION >= 19)
```
[the remark this answers is not in the document]
Comment (caoxiaoyuyuyuyuyu) [reply csrc/kernels/legacy/ibgda_device.cuh:89]: Addressed in 2c70ee1. DeepEP now calls `nvshmemx_vendor_get_version_info` before `nvshmemx_init_attr` and rejects compile/runtime pairs when their RC-QP layout classes differ across the 3.5.19 transition. The error reports both versions and asks the user to rebuild against the loaded library, so no QPs are initialized under an incompatible layout.

[the remark this answers is not in the document]
Comment (caoxiaoyuyuyuyuyu) [reply csrc/kernels/legacy/ibgda_device.cuh:89]: Additional source confirmation: `nvshmemx_vendor_get_version_info` is a public host API and returns the vendor macros compiled into the loaded host library itself. It is declared and implemented in each checked supported release: [3.4.5 declaration](https://github.com/NVIDIA/nvshmem/blob/v3.4.5-0/src/include/host/nvshmem_api.h#L78-L79) / [implementation](https://github.com/NVIDIA/nvshmem/blob/v3.4.5-0/src/host/init/query_host.cpp#L34-L38), [3.5.19 declaration](https://github.com/NVIDIA/nvshmem/blob/v3.5.19-1/src/include/host/nvshmem_api.h#L81-L82) / [implementation](https://github.com/NVIDIA/nvshmem/blob/v3.5.19-1/src/host/init/query_host.cpp#L34-L38), and [3.7.2 declaration](https://github.com/NVIDIA/nvshmem/blob/v3.7.2-0/src/include/host/nvshmem_api.h#L74-L75) / [implementation](https://github.com/NVIDIA/nvshmem/blob/v3.7.2-0/src/host/init/query_host.cpp#L34-L38). The new check therefore observes the loaded library version rather than the compile-time headers.

**Comment by alpha-baby** [comment]
same issue https://github.com/deepseek-ai/DeepEP/issues/707

File: csrc/kernels/backend/nvshmem.cu [hunk 4]
```diff
@@ -1,17 +1,49 @@
-#include <cstring>
-#include <optional>
-#include <vector>
+#include <nvshmem.h>
```
Comment (MARD1NO) [inline csrc/kernels/backend/nvshmem.cu:1]: why import nvshmem.h before compiled.cuh, it will cause 

```
/nvidia/nvshmem/include/non_abi/device/coll/reduce.
 cuh(101): note #3328-D: built-in operator<=>(<nullptr>, <nullptr>) does not match because argument #1 does not match
 parameter
               return (op1 > op2) ? op1 : op2;
```

Comment (alpha-baby) [reply csrc/kernels/backend/nvshmem.cu:1]: maybe you could try this PR: https://github.com/deepseek-ai/DeepEP/pull/564
